### PR TITLE
[Speedy] add cooperative interuptions

### LINKED ISSUE #13954: Ensure scenario service finishes before gRPC deadline
Currently, if the scenario service is given a script that loops, for example,

```
module Main where

import Daml.Script

main : Script ()
main = do
  pure ()
  main
```

daml studio will (eventually) only show the error message 

```
"Scenario service backend error: BErrorClient (ClientIOError (GRPCIOBadStatusCode StatusDeadlineExceeded (StatusDetails {unStatusDetails = \"Deadline Exceeded\"})))"
```

This is because the client stops waiting after a default of 60 seconds, while the server gets stuck in an infinite loop. Instead, the server should check the deadline for each request and `await` on the result of the request only for the duration of the deadline (perhaps subtracting a few seconds to be safe).

We should try to turn this into a proper script execution failure, if possible showing the partial ledger state before the timeout. 

Note: this was previously https://github.com/digital-asset/daml/issues/13518, but the associated PR fixes other fatal errors instead.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Part of #13954

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
Benchmark do not detect any regressions:

main (39d8845b)
```
Benchmark                                                             (dar)             (scenario)  Mode  Cnt   Score   Error  Units
CollectAuthority.bench  //daml-lf/scenario-interpreter/CollectAuthority.dar  CollectAuthority:test  avgt  200  11.856 ± 0.027  ms/op
```
remy-speedy (b627a803)
```
Benchmark                                                             (dar)             (scenario)  Mode  Cnt   Score   Error  Units
CollectAuthority.bench  //daml-lf/scenario-interpreter/CollectAuthority.dar  CollectAuthority:test  avgt  200  11.783 ± 0.056  ms/op
```

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala [hunk 1]
```diff
@@ -74,6 +74,9 @@ private[lf] object Pretty {
       case Error.PartiesNotAllocated(parties) =>
         text(s"Error: Tried to submit a command for parties that have not been allocated:") &
           intercalate(comma + space, parties.map(prettyParty))
+
+      case Error.Timeout(timeout) =>
+        text(s"Timeout: evaluation needed more that ${timeout.toSeconds}s. to complete")
```
Comment (akrmn) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala:79]:
```suggestion
        text(s"Timeout: evaluation needed more that ${timeout.toSeconds}s to complete")
```

File: daml-lf/tests/scenario/stable/timeout/EXPECTED.ledger [hunk 2]
```diff
@@ -0,0 +1 @@
+Timeout: evaluation needed more that 5s. to complete
```
Comment (akrmn) [inline daml-lf/tests/scenario/stable/timeout/EXPECTED.ledger:1]:
```suggestion
Timeout: evaluation needed more that 5s to complete
```

File: daml-lf/tests/scenario/stable/timeout/Test.daml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Test where
+
+loop 0 = ()
+loop x = loop (x + 1)
```
Comment (akrmn) [inline daml-lf/tests/scenario/stable/timeout/Test.daml:7]: I don't think you need the first clause other than for the type, so

```suggestion
loop : Int -> ()
loop x = loop (x + 1) 
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 4]
```diff
@@ -450,6 +450,7 @@ private[lf] object Speedy {
         committers: Set[Party],
         readAs: Set[Party],
         authorizationChecker: AuthorizationChecker = DefaultAuthorizationChecker,
+        iterationsBetweenInterruptions: Long = 10000,
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:453]:
```suggestion
        iterationsBetweenInterruptions: Long = ScenarioMachine.defaultIterationsBetweenInterruptions,
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 5]
```diff
@@ -92,14 +92,18 @@ object Question {
...
 
   /** The speedy machine has completed evaluation to reach a final value.
     * And, if the evaluation was on-ledger, a completed transaction.
     */
   final case class SResultFinal(v: SValue) extends SResult[Nothing]
 
-  final case class SResultQuestion[Q](question: Q) extends SResult[Q]
+  final case class SResultError(err: SError) extends SResult[Nothing]
+
+  final case class SResultInterruption(iterations: Long, callback: () => Unit)
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:105]: `iterations` is never used, I think we can drop it (and thus can also drop `steps` from `daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala`)

**Review by akrmn — APPROVED** [review]
LGTM, just a few minor suggestions. Thank you!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 6]
```diff
@@ -92,14 +92,17 @@ object Question {
 sealed abstract class SResult[+Q] extends Product with Serializable
 
 object SResult {
-  final case class SResultError(err: SError) extends SResult[Nothing]
+
+  final case class SResultQuestion[Q](question: Q) extends SResult[Q]
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:96]: Just a small note/complaint. I would prefer code cleanups, reordering and other small refactorings to be in different PRs, as it makes the substantive changes harder to review.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 7]
```diff
@@ -92,14 +92,17 @@ object Question {
...
 
   /** The speedy machine has completed evaluation to reach a final value.
     * And, if the evaluation was on-ledger, a completed transaction.
     */
   final case class SResultFinal(v: SValue) extends SResult[Nothing]
 
-  final case class SResultQuestion[Q](question: Q) extends SResult[Q]
+  final case class SResultError(err: SError) extends SResult[Nothing]
+
+  final case class SResultInterruption(callback: () => Unit) extends SResult[Nothing]
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:105]: I think the callback argument can just be removed.
It appears that every client which matches will just immediately call it before re-invoking run()

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 8]
```diff
@@ -816,32 +827,33 @@ private[lf] object Speedy {
...
-            case Control.Question(res) =>
-              SResultQuestion(res)
-            case Control.Complete(value: SValue) =>
-              if (enableInstrumentation) track.print()
-              SResultFinal(value)
-            case Control.Error(ie) =>
-              SResultError(SErrorDamlException(ie))
+          if (interruptionCountDown == 0) {
+            interruptionCountDown = iterationsBetweenInterruptions
+            SResultInterruption(() => control = thisControl)
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:836]: Remove the callback parameter from SResultInterruption.
Move the assignment to `control` to just after the assignment to `interruptionCountdown`.

**Review by nickchapman-da — APPROVED** [review]
Looks good to me.
Small simplification suggested.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:96]: :+1:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala]: :+1:

**Pull request merged**